### DISCUSSION TITLE
mpremote rfc2217 telnet connection (rp2040 / Raspberry Pi Pico W)

### DISCUSSION THREAD

**Discussion by scruss** [micropython/micropython#14308] [Using MicroPython]
[mpremote connect](https://docs.micropython.org/en/latest/reference/mpremote.html#mpremote-command-connect) supports communication over [RFC 2217 - Telnet Com Port Control Option](https://datatracker.ietf.org/doc/html/rfc2217). What MicroPython code would I need to be running on a Raspberry Pi Pico W to make this remote connection work, please?

Attempts to use [utelnetserver](https://github.com/cpopp/MicroTelnetServer) result in `failed to access rfc2217://192.168.2.126:23 (it may be in use by another program)`. If I telnet to the same address using the `telnet` command, though, I get access to the REPL.

**Comment by Josverl** [comment]
@scruss,  I added that option to support connection to Wokwi that can emulate espxx and rp2 boards in  #12500 
since i moved off the LoBo ports, i have missed a stable and performant telnet server (on esp32) for remote connections. 
I've doe a few tests with telnet servers in MicroPython - but not to my satisfaction.

if you want a direct mpremote connection , it may be possible to just skip Telnet altogether as I mentioned in that PR:
_'It would be simple to add support for a straight tcp pipe ( using socket://host:port ) if there is use for that as well.'_
I think we just need the device to listen on a known socket , wrap the socket in a stream / IOBase  and then dupterm to that stream. 

conceptually something like 
``` python 
## completely untested code ##
s = usocket.socket(usocket.AF_INET, usocket.SOCK_STREAM)
s.bind(('0.0.0.0', 4321))
s.listen(1)
while True:
  (clientsocket, address) = s.accept()
  # now do something with the clientsocket
  # pretend this is a streamed filelike object
  stream = uio.IOBase(clientsocket)
  os.dupterm(stream)
## completely untested code ##
```
need some safeguards and error handling around that , and security will be non-existent 

a small mpremote/transport_serial.py patch would be needed to also check for `socket://`
```
                if device.startswith("rfc2217://", "socket://"):
                    self.serial = serial.serial_for_url(device, **serial_kwargs)

```

**Reply by scruss** [reply]
Ah, thanks. I suspect I'm misunderstanding the intended usage of a rfc2217 connection. I'd hoped it was an alternative to the missing WebREPL support in mpremote

**Reply by Josverl** [reply]
Hear hear, I'd love to have that work with a telnet server on the device too

**Reply by andrewleech** [reply]
mpremote webrepl support has been proposed in https://github.com/micropython/micropython/pull/14141 using the webrepl websocket.
